### PR TITLE
Pool busy event emitting on all pool types

### PR DESCRIPTION
**Pull request by jerome-benoit**
Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

## PR Checklist

- [x] Please add a description of your changes.
- [x] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [ ] Please add a link to the open issue or task that this pull request will resolve.
- [ ] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

### CODE REVIEW

File: src/pools/abstract-pool.ts [hunk 1]
```diff
@@ -222,6 +225,27 @@ export abstract class AbstractPool<
...
+    if (this.dynamic && this.workers.length === this.max) {
+      return true
+    } else if (this.findFreeTasksMapEntry() === [null, null]) {
+      return true
+    }
+    return false
+  }
+
+  /** @inheritdoc */
+  public findFreeTasksMapEntry (): [Worker, number] | [null, null] {
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:239]: I would prefer to return something like `undefined`, `null` or `false`  
So e.g.
```suggestion
  public findFreeTasksMapEntry (): [Worker, number] | false {
```

File: src/pools/abstract-pool.ts [hunk 2]
```diff
@@ -222,6 +225,27 @@ export abstract class AbstractPool<
...
+    for (const [worker, numberOfTasks] of this.tasks) {
+      if (numberOfTasks === 0) {
+        // A worker is free, return the matching tasks map entry
+        return [worker, numberOfTasks]
+      }
+    }
+    return [null, null]
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:246]:
```suggestion
    return false
```

File: src/pools/abstract-pool.ts [hunk 3]
```diff
@@ -399,4 +424,10 @@ export abstract class AbstractPool<
       }
     }
   }
+
+  private checkEmitBusy (): void {
+    if (this.isPoolBusy()) {
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:429]: Do you thing that we should use a getter also for `isPoolBusy`? Like as for `isDynamic`?

File: src/pools/pool-internal.ts [hunk 4]
```diff
@@ -52,4 +52,22 @@ export interface IPoolInternal<
...
+  /**
+   * Find a tasks map entry with a free worker based on the number of tasks the worker has applied.
+   *
+   * If a worker was found that has `0` tasks, it is detected as free.
+   *
+   * If no tasks map entry with a free worker was found, `null` will be returned.
+   *
+   * @returns A tasks map entry with a free worker if there was one, otherwise `null`.
+   */
+  findFreeTasksMapEntry(): [Worker, number] | [null, null]
```
Comment (Shinigami92) [inline src/pools/pool-internal.ts:72]:
```suggestion
  findFreeTasksMapEntry(): [Worker, number] | false
```

File: src/pools/pool-internal.ts [hunk 5]
```diff
@@ -52,4 +52,22 @@ export interface IPoolInternal<
...
+  isPoolBusy(): boolean
+
+  /**
+   * Find a tasks map entry with a free worker based on the number of tasks the worker has applied.
+   *
+   * If a worker was found that has `0` tasks, it is detected as free.
+   *
+   * If no tasks map entry with a free worker was found, `null` will be returned.
+   *
+   * @returns A tasks map entry with a free worker if there was one, otherwise `null`.
```
Comment (Shinigami92) [inline src/pools/pool-internal.ts:70]: Do not forget to update the JSDoc

File: src/pools/selection-strategies.ts [hunk 6]
```diff
@@ -136,15 +136,12 @@ class DynamicPoolWorkerChoiceStrategy<Worker extends IWorker, Data, Response>
 
   /** @inheritdoc */
   public choose (): Worker {
-    const freeWorker = SelectionStrategiesUtils.findFreeWorkerBasedOnTasks(
-      this.pool.tasks
-    )
+    const freeWorker = this.pool.findFreeTasksMapEntry()[0]
```
Comment (Shinigami92) [inline src/pools/selection-strategies.ts:139]: Mh :thinking:  
Maybe `null` instead of `false` would be better, I'm not sure

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

**Comment by pioardi** [comment]
@jerome-benoit  @Shinigami92  changing FullPool to busy is a breaking change so if we release now the version 2.x.x we would release a 3.x.x if we include this change....
There is no bads to increase from 2 to 3 for a small breaking change but you know...I don't like that too much so please let me know if you want to do that before we release the version 2....if not, we will need to hold on this until we don't have other major changes else does not make too much sense to me to have a version 3 just for this little change

**Comment by Shinigami92** [comment]
@jerome-benoit Hurry up :smile: Don't make @pioardi wait XD

@pioardi Optionally you could rely on humans and just don't force push the release today or tomorrow

**Comment by pioardi** [comment]
> @jerome-benoit Hurry up 😄 Don't make @pioardi wait XD
> 
> @pioardi Optionally you could rely on humans and just don't force push the release today or tomorrow

Lol we can work on this before to release, It Is not a problem :)

File: src/pools/abstract-pool.ts [hunk 7]
```diff
@@ -311,6 +335,7 @@ export abstract class AbstractPool<
    * @returns Worker.
    */
   protected chooseWorker (): Worker {
+    this.checkEmitBusy()
```
Comment (pioardi) [inline src/pools/abstract-pool.ts:338]: The entire checkEmitBusy method could have a performance overhead IMO.
We should at least have an implementation for each type of pool ( static,dynamic ).
One thing that I would consider instead is to check the number of current tasks ( promiseMap.size() ) and if is greater than the pool size ( min or max parameters ) than I would emit an event ( to discuss more details on that).
Another thing that I would do is define a maxConcurrentTasks option and then check if this number is passed, if yes then we can emit another event here ( still using the promiseMap.size(), and also take optional backpressure actions on it directly on poolifier ( i.e. backpressure delay, delay each task to X ms later without any impact on the user that still receive a promise ) .

**Review by pioardi — CHANGES_REQUESTED** [review]
I tried to collect my thoughts into one comment

**Comment by pioardi** [comment]
@Shinigami92  @jerome-benoit  good morning everyone.
So this is what I was thinking to do, we can just rename the event FullPool to fullpool and then we will keep it later on.
All the new stuff will be new so we will not break compatibility for tiny things.

I will hold on for some hours and then I will proceed with this rename :) So we can work on new things quietly and later on :)

**Comment by Shinigami92** [comment]
> @Shinigami92 @jerome-benoit good morning everyone.
> So this is what I was thinking to do, we can just rename the event FullPool to fullpool and then we will keep it later on.
> All the new stuff will be new so we will not break compatibility for tiny things.
> 
> I will hold on for some hours and then I will proceed with this rename :) So we can work on new things quietly and later on :)

I would prefer the `'busy'` wording because it makes much more sense

I'm currently very busy (ha ^^ pun not intended) cause of the JSWorld Conference 2021 and specially today and tomorrows Vue Amsterdam Conference

**Comment by pioardi** [comment]
> > @Shinigami92 @jerome-benoit good morning everyone.
> > So this is what I was thinking to do, we can just rename the event FullPool to fullpool and then we will keep it later on.
> > All the new stuff will be new so we will not break compatibility for tiny things.
> > I will hold on for some hours and then I will proceed with this rename :) So we can work on new things quietly and later on :)
> 
> I would prefer the `'busy'` wording because it makes much more sense
> 
> I'm currently very busy (ha ^^ pun not intended) cause of the JSWorld Conference 2021 and specially today and tomorrows Vue Amsterdam Conference

Beautiful events :) I would say the better fit could be...maxpoolsize ( much more near to the real meaning of this event ) what you think bro ?

**Comment by Shinigami92** [comment]
> Beautiful events :) I would say the better fit could be...maxpoolsize ( much more near to the real meaning of this event ) what you think bro ?

I prefer real-short event names, best case would be just one word
`maxpoolsize` is build with three words
`busy` fulfills everything what I prefer

**Comment by pioardi** [comment]
> > Beautiful events :) I would say the better fit could be...maxpoolsize ( much more near to the real meaning of this event ) what you think bro ?
> 
> I prefer real-short event names, best case would be just one word
> `maxpoolsize` is build with three words
> `busy` fulfills everything what I prefer

mmmmm make sense....but you know busy means busy....is not the real meaning of this event...this event means that the max size of the pool is reached.....mmmmmm

**Comment by jerome-benoit** [comment]
> > > Beautiful events :) I would say the better fit could be...maxpoolsize ( much more near to the real meaning of this event ) what you think bro ?
> > 
> > 
> > I prefer real-short event names, best case would be just one word
> > `maxpoolsize` is build with three words
> > `busy` fulfills everything what I prefer
> 
> mmmmm make sense....but you know busy means busy....is not the real meaning of this event...this event means that the max size of the pool is reached.....mmmmmm

Nope, it means: number of workers in the pool have reached the pool maximum size **AND** that all workers are busy. So naming that event 'busy' for all pool type (existing and future) has a lot of sense.

**Comment by pioardi** [comment]
> > > > Beautiful events :) I would say the better fit could be...maxpoolsize ( much more near to the real meaning of this event ) what you think bro ?
> > > 
> > > 
> > > I prefer real-short event names, best case would be just one word
> > > `maxpoolsize` is build with three words
> > > `busy` fulfills everything what I prefer
> > 
> > 
> > mmmmm make sense....but you know busy means busy....is not the real meaning of this event...this event means that the max size of the pool is reached.....mmmmmm
> 
> Nope, it means: number of workers in the pool have reached the pool maximum size **AND** that all workers are busy. So naming that event 'busy' for all pool types (existing and future) has a lot of sense.

I Will proceed with busy :) for the others changes we Will discuss when Is time then :)

**Comment by pioardi** [comment]
FullPool renamed to busy directly on master, change included to the changelog.

File: src/pools/selection-strategies.ts (hunk 6 above)
Comment (jerome-benoit) [reply src/pools/selection-strategies.ts:139]: You can't return a restructured array and a boolean, the code need to access each member of the array.

File: tests/pools/cluster/dynamic.test.js [hunk 8]
```diff
@@ -21,14 +21,14 @@ describe('Dynamic cluster pool test suite', () => {
...
-    pool.emitter.on('busy', () => busy++)
+    let poolBusy = 0
+    pool.emitter.on('busy', () => poolBusy++)
     for (let i = 0; i < max * 2; i++) {
       promises.push(pool.execute({ test: 'test' }))
     }
     expect(pool.workers.length).toBeLessThanOrEqual(max)
     expect(pool.workers.length).toBeGreaterThan(min)
-    expect(busy > 1).toBeTruthy()
+    expect(poolBusy).toEqual(max + 1)
```
Comment (jerome-benoit) [inline tests/pools/cluster/dynamic.test.js:31]: I think the code emit the busy event one time more on dynamic pool.

File: src/pools/selection-strategies.ts (hunk 6 above)
Comment (Shinigami92) [reply src/pools/selection-strategies.ts:140]: Please check my PR #243

File: src/pools/pool-internal.ts [hunk 9]
```diff
@@ -2,6 +2,14 @@ import EventEmitter from 'events'
 import type { IWorker } from './abstract-pool'
 import type { IPool } from './pool'
 
+/**
+ * Pool types.
+ */
+export enum PoolType {
```
Comment (Shinigami92) [inline src/pools/pool-internal.ts:8]: As long as we do not export this to public, this can be an TS enum.  
Otherwise we should use `Object.freeze` as the other _enums_.

File: src/pools/abstract-pool.ts [hunk 10]
```diff
@@ -399,4 +415,10 @@ export abstract class AbstractPool<
       }
     }
   }
+
+  private checkEmitBusy (): void {
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:419]: I don't know why, but I don't like the function's name (`checkEmitBusy`) :confused:  
Maybe `checkAndEmitBusy`? Other suggestions?

**Comment by Shinigami92** [comment]
@pioardi From my side this PR is now ready to get merged :slightly_smiling_face:  
When you approve it, @jerome-benoit can merge it and see my review as an approval too

/cc @jerome-benoit

**Comment by jerome-benoit** [comment]
> @pioardi From my side this PR is now ready to get merged 🙂
> When you approve it, @jerome-benoit can merge it and see my review as an approval too

I think the code still have an issue at emitting the right number of busy events on fixed pool. 
On dynamic pool, the test emit if all workers run a task: so number of iteration + 1. 
On fixed pool, the busy event counting is not correct yet, I think the event is not emitted when the numberOfWorkers all have a task exactly.

**Comment by jerome-benoit** [comment]
> > @pioardi From my side this PR is now ready to get merged 🙂
> > When you approve it, @jerome-benoit can merge it and see my review as an approval too
> 
> I think the code still have an issue at emitting the right number of busy events on fixed pool.
> On dynamic pool, the test emit if all workers run a task: so number of iteration + 1.
> On fixed pool, the busy event counting is not correct yet, I think the event is not emitted when the numberOfWorkers workers in pool all have one task exactly.

Or it's the other way around but the code have an issue.

**Comment by pioardi** [comment]
A couple of hours and I Will take a look

File: tests/pools/cluster/dynamic.test.js (hunk 8 above)
Comment (pioardi) [inline tests/pools/cluster/dynamic.test.js:31]: So this is a bug ?

File: src/pools/abstract-pool.ts [hunk 11]
```diff
@@ -222,11 +223,33 @@ export abstract class AbstractPool<
     )
   }
 
+  /** @inheritdoc */
+  public abstract get busy (): boolean
+
+  protected internalGetBusyStatus (): boolean {
+    return (
+      this.promiseMap.size >= this.numberOfWorkers &&
```
Comment (pioardi) [inline src/pools/abstract-pool.ts:231]: >== would be better ?

Comment (pioardi) [reply src/pools/abstract-pool.ts:231]: > ==

**Comment by jerome-benoit** [comment]
> 
> In tests/pools/cluster/dynamic.test.js <https://github.com/pioardi/poolifier/pull/241#discussion_r584265494>:
> 
> >      for (let i = 0; i < max * 2; i++) {
>        promises.push(pool.execute({ test: 'test' }))
>      }
>      expect(pool.workers.length).toBeLessThanOrEqual(max)
>      expect(pool.workers.length).toBeGreaterThan(min)
> -    expect(busy > 1).toBeTruthy()
> +    expect(poolBusy).toEqual(max + 1)
> So this is a bug ?
> 
> 
> 

I think so. The loop must only trigger max times the event, like for fixed pool. And I do not find the reason for it.

--
Jérôme Benoit aka fraggle
Piment Noir - https://piment-noir.org
OpenPGP Key ID : 27B535D3
Key fingerprint : B799 BBF6 8EC8 911B B8D7 CDBC C3B1 92C6 27B5 35D3

File: src/pools/thread/fixed.ts [hunk 12]
```diff
@@ -81,4 +82,14 @@ export class FixedThreadPool<
...
   }
+
+  /** @inheritdoc */
+  public get type (): PoolType {
+    return PoolType.FIXED
+  }
+
+  /** @inheritdoc */
+  public get busy (): boolean {
+    return this.internalGetBusyStatus()
```
Comment (pioardi) [inline src/pools/thread/fixed.ts:93]: So the performance overhead will be mainly on the fixed pools right ?  I was thinking if is really useful for end users to know if all the workers are busy ? What they can do in that case I mean ?

**Review by pioardi — CHANGES_REQUESTED** [review]
Consider the performance overhead of adding this event to the fixed pools.
Consider which kind of actions can take an end users in the case on a busy event.

File: src/pools/thread/fixed.ts (hunk 12 above)
Comment (jerome-benoit) [reply src/pools/thread/fixed.ts:93]: The job admission can't currently be regulated by users. I, for example, use a homemade queue at job admission and I need to trigger it only if needed.

File: src/pools/abstract-pool.ts (hunk 11 above)
Comment (jerome-benoit) [reply src/pools/abstract-pool.ts:231]: The second condition needs to be triggered only if the number of tasks is superior or equal to the number of workers.

**Comment by pioardi** [comment]
> Consider the performance overhead of adding this event to the fixed pools.
> Consider which kind of actions can take an end users in the case on a busy event.

My answer to the second question is that a user could apply back pressure logic, but in my idea we should have another option maxConcurrentTasks ( or maxTasksInProgress ) that we can easily check using the promiseMap size, in that case users can do back pressure logic and we do not have any performance overhead, moreover we can also implement a built-in back pressure like back pressure delay .

**Comment by jerome-benoit** [comment]
> > Consider the performance overhead of adding this event to the fixed pools.
> > Consider which kind of actions can take an end users in the case on a busy event.
> 
> My answer to the second question is that a user could apply back pressure logic, but in my idea we should have another option maxConcurrentTasks ( or maxTasksInProgress ) that we can easily check using the promiseMap size, in that case users can do back pressure logic and we do not have any performance overhead, moreover we can also implement a built-in back pressure like back pressure delay .

A busy pool and the number of tasks running are too different things. 
The first is used to regulate the tasks admission, the second to regulate the ressources taken.

File: src/pools/abstract-pool.ts [hunk 13]
```diff
@@ -87,6 +87,10 @@ export interface PoolOptions<Worker> {
    * The work choice strategy to use in this pool.
    */
   workerChoiceStrategy?: WorkerChoiceStrategy
+  /**
+   * Pool events emitting.
+   */
+  emitEvents?: boolean
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:93]: From the name of this variable, it's not clear to me what it's doing.  
Can you try to find a better name?

File: src/pools/pool-internal.ts [hunk 14]
```diff
@@ -39,17 +47,40 @@ export interface IPoolInternal<
...
+
+  /**
+   * Whether the pool is busy or not.
+   *
+   * The pool busyness boolean status.
+   */
+  busy: boolean
+
+  /**
+   * Number of tasks running concurrently.
```
Comment (Shinigami92) [inline src/pools/pool-internal.ts:72]: Currently or max?

Comment (Shinigami92) [reply src/pools/pool-internal.ts]: @jerome-benoit MUCH better :+1:

**Comment by pioardi** [comment]
I like the idea ti have an option for event emissions

**Comment by Shinigami92** [comment]
:thinking: I think the coverage dropped a bit
Maybe cause of this line?:
https://sonarcloud.io/code?id=pioardi_poolifier&pullRequest=241&selected=pioardi_poolifier%3Asrc%2Fpools%2Fabstract-pool.ts&line=443

**Comment by jerome-benoit** [comment]
I think it's now in a mergeable state: 
+ I do not know how to cover sonarcloud.io/code?id=pioardi_poolifier&pullRequest=241&selected=pioardi_poolifier%3Asrc%2Fpools%2Fabstract-pool.ts&line=443, I've added a test that check that emitter attribute is defined or not with no luck
+ After the PR merge, `npm run typedoc` shall be run by @pioardi 
+ An issue shall be created regarding the busy event count on dynamic pool.

File: src/pools/abstract-pool.ts [hunk 15]
```diff
@@ -156,13 +165,14 @@ export abstract class AbstractPool<
...
     this.checkFilePath(this.filePath)
+    this.checkPoolOptions(opts)
     this.setupHook()
 
     for (let i = 1; i <= this.numberOfWorkers; i++) {
       this.createAndSetupWorker()
     }
 
-    this.emitter = new PoolEmitter()
+    this.opts.enableEvents && (this.emitter = new PoolEmitter())
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:175]:
```suggestion
    if (this.opts.enableEvents) {
      this.emitter = new PoolEmitter()
    }
```

File: src/pools/abstract-pool.ts [hunk 16]
```diff
@@ -179,7 +189,7 @@ export abstract class AbstractPool<
         })
         return workerCreated
       },
-      opts.workerChoiceStrategy ?? WorkerChoiceStrategies.ROUND_ROBIN
+      opts?.workerChoiceStrategy ?? WorkerChoiceStrategies.ROUND_ROBIN
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:192]: If I see it correctly, we don't need this optional chain anymore :thinking:  
Because on line 175 we already access it without any check
```suggestion
      opts.workerChoiceStrategy ?? WorkerChoiceStrategies.ROUND_ROBIN
```

File: src/pools/abstract-pool.ts [hunk 17]
```diff
@@ -202,14 +212,22 @@ export abstract class AbstractPool<
...
       )
-    } else if (!this.dynamic && numberOfWorkers === 0) {
+    } else if (this.type === PoolType.FIXED && numberOfWorkers === 0) {
       throw new Error('Cannot instantiate a fixed pool with no worker')
     }
   }
 
+  private checkPoolOptions (opts?: PoolOptions<Worker>): void {
+    this.opts.enableEvents =
+      opts?.enableEvents !== undefined ? opts.enableEvents : true
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:222]: What if `enableEvents` would be `null`? Theoretically possible if ignoring TS while using JS

File: tests/pools/cluster/fixed.test.js [hunk 18]
```diff
@@ -87,7 +98,7 @@ describe('Fixed cluster pool test suite', () => {
       inError = e
     }
     expect(inError).toBeDefined()
-    expect(typeof inError === 'string').toBeTruthy()
+    expect(typeof inError === 'string').toEqual(true)
```
Comment (Shinigami92) [inline tests/pools/cluster/fixed.test.js:101]: Why not `.toBeTruthy()`? I think we can safely use it here

File: tests/pools/cluster/fixed.test.js [hunk 19]
```diff
@@ -100,7 +111,7 @@ describe('Fixed cluster pool test suite', () => {
...
```
Comment (Shinigami92) [inline tests/pools/cluster/fixed.test.js:114]: Why not `.toBeTruthy()`? I think we can safely use it here

File: tests/pools/thread/fixed.test.js [hunk 20]
```diff
@@ -80,9 +90,10 @@ describe('Fixed thread pool test suite', () => {
     } catch (e) {
       inError = e
     }
-    expect(inError).toBeTruthy()
-    expect(inError instanceof Error).toBeTruthy()
-    expect(inError.message).toBeTruthy()
+    expect(inError).toBeDefined()
+    expect(inError).toBeInstanceOf(Error)
+    expect(inError.message).toBeDefined()
+    expect(typeof inError.message === 'string').toEqual(true)
```
Comment (Shinigami92) [inline tests/pools/thread/fixed.test.js:96]: Why not `.toBeTruthy()`? I think we can safely use it here

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

File: tests/pools/cluster/fixed.test.js (hunk 18 above)
Comment (jerome-benoit) [reply tests/pools/cluster/fixed.test.js:101]: Stricter returned value check: typeof return a boolean, never null, '', etc.

File: tests/pools/cluster/fixed.test.js (hunk 19 above)
Comment (jerome-benoit) [reply tests/pools/cluster/fixed.test.js:114]: Stricter returned value check: typeof return a boolean, never null, '', etc.

File: tests/pools/thread/fixed.test.js (hunk 20 above)
Comment (jerome-benoit) [reply tests/pools/thread/fixed.test.js:96]: Stricter returned value check: typeof return a boolean, never null, '', etc.

File: src/pools/abstract-pool.ts [hunk 21]
```diff
@@ -202,14 +214,24 @@ export abstract class AbstractPool<
...
       )
-    } else if (!this.dynamic && numberOfWorkers === 0) {
+    } else if (this.type === PoolType.FIXED && numberOfWorkers === 0) {
       throw new Error('Cannot instantiate a fixed pool with no worker')
     }
   }
 
+  private checkPoolOptions (opts: PoolOptions<Worker>): void {
+    this.opts.enableEvents =
+      opts.enableEvents !== undefined || opts?.enableEvents !== null
```
Comment (Shinigami92) [inline src/pools/abstract-pool.ts:224]: This can be simplified to `opts.enableEvents != null`

Comment (Shinigami92) [reply src/pools/abstract-pool.ts:224]: Or I think it also can be shortened to `this.opts.enableEvents = opts.enableEvents ?? true`

File: tests/pools/thread/fixed.test.js (hunk 20 above)
Comment (Shinigami92) [reply tests/pools/thread/fixed.test.js:96]: But a `typeof ... ===` check can only return true or false

Comment (jerome-benoit) [reply tests/pools/thread/fixed.test.js:96]: > But a `typeof ... ===` check can only return true or false

So tests now strictly check for a boolean 'true' value now ;)

File: src/pools/abstract-pool.ts (hunk 21 above)
Comment (jerome-benoit) [reply src/pools/abstract-pool.ts]: The goal here is the check and affect the value given at objet instantiation.

Comment (Shinigami92) [reply src/pools/abstract-pool.ts]: Yeah I know, it the same but just smaller syntax
The `??` operator is really new to JS, you should read about optional coalescing
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

**Comment by pioardi** [comment]
@jerome-benoit @Shinigami92  going to release today, tomorrow we will share across all the channels.
I will review this PR as soon as I can post release and share

**Comment by jerome-benoit** [comment]
> @jerome-benoit @Shinigami92 going to release today, tomorrow we will share across all the channels.
> I will review this PR as soon as I can post release and share

Well, you should release after the merge of that PR or integrate from it in master some of the missed code path related to event renaming. 
The post-release check list requires that you generate the docs content.

**Comment by pioardi** [comment]
> > @jerome-benoit @Shinigami92 going to release today, tomorrow we will share across all the channels.
> > I will review this PR as soon as I can post release and share
> 
> Well, you should release after the merge of that PR or integrate from it in master some of the missed code path related to event renaming.
> The post-release check list requires that you generate the docs content.

Event is already renamed on the master branch from FullPool to busy, any other change needed ?

**Comment by jerome-benoit** [comment]
> Well, you should release after the merge of that PR or integrate from it in master some of the missed code path related to event renaming.
> The post-release check list requires that you generate the docs content.
> 
> Event is already renamed on the master branch from FullPool to busy, any other change needed ?
> 

Like I said: code paths have been missed, look code changes in that PR or search for the full keyword in the code.

--
Jérôme Benoit aka fraggle
Piment Noir - https://piment-noir.org
OpenPGP Key ID : 27B535D3
Key fingerprint : B799 BBF6 8EC8 911B B8D7 CDBC C3B1 92C6 27B5 35D3

**Comment by pioardi** [comment]
> Well, you should release after the merge of that PR or integrate from it in master some of the missed code path related to event renaming. The post-release check list requires that you generate the docs content. Event is already renamed on the master branch from FullPool to busy, any other change needed ?
> Like I said: code paths have been missed, look code changes in that PR or search for the full keyword in the code.
> […](#)

The FullPool key is no more present on master branch, maybe some logs are still saying "the pool is full" but is not so essential to change it before release, anyway since that is fast I now did a commit to change them and hopefully I fixed the problem with the GH actions

**Comment by jerome-benoit** [comment]
> Like I said: code paths have been missed, look code changes in that PR or search for the full keyword in the code.
> … <x-msg://2/#>
> The FullPool key is no more present on master branch, maybe some logs are still saying "the pool is full" but is not so essential to change it before release, anyway since that is fast I now did a commit to change them and hopefully I fixed the problem with the GH actions
> 


I’ve written: 'search for the full keyword in the code’, not fullPool ;)

--
Jérôme Benoit aka fraggle
Piment Noir - https://piment-noir.org
OpenPGP Key ID : 27B535D3
Key fingerprint : B799 BBF6 8EC8 911B B8D7 CDBC C3B1 92C6 27B5 35D3

**Comment by pioardi** [comment]
> Like I said: code paths have been missed, look code changes in that PR or search for the full keyword in the code. … <x-msg://2/#> The FullPool key is no more present on master branch, maybe some logs are still saying "the pool is full" but is not so essential to change it before release, anyway since that is fast I now did a commit to change them and hopefully I fixed the problem with the GH actions
> I’ve written: 'search for the full keyword in the code’, not fullPool ;)
> […](#)

Oh yep I did it, I don 't know why now GH actions don't start... any idea why ?

**Comment by Shinigami92** [comment]
Do you want to merge this and release a v2.1.0 or v2.0.1 or so?

**Comment by pioardi** [comment]
> Do you want to merge this and release a v2.1.0 or v2.0.1 or so?

I still need to review the PR, once than we can merge and release a 2.0.1.
A couple of days time that we share the new version and stuff and we will go through this also =)

**Comment by jerome-benoit** [comment]
> > Do you want to merge this and release a v2.1.0 or v2.0.1 or so?
> 
> I still need to review the PR, once than we can merge and release a 2.0.1.
> A couple of days time that we share the new version and stuff and we will go through this also =)

I do not see what you still need to review here. That PR just make the event emitter API consistent across pool types. By doing so, it makes the code more separated into more logical parts and cleaner. The tiny overhead added on fixed pool is exactly the same as the one dynamic pool had without impacting their performance in real world usage, and it can even be disabled if needed.

File: src/pools/pool-internal.ts (hunk 9 above)
Comment (pioardi) [reply src/pools/pool-internal.ts:8]: Any chance to use Object.freeze here?

**Review by pioardi — CHANGES_REQUESTED** [review]
Just one small comment and we are ready to go

File: src/pools/pool-internal.ts (hunk 9 above)
Comment (jerome-benoit) [reply src/pools/pool-internal.ts:8]: I've not really well understood why @Shinigami92 changed TS enum in the first place. 
The enum transpiled code looks really well optimised to me with TS 4.2.x, so I see no gain on using `const`, `Object.freeze`, etc.

Comment (Shinigami92) [reply src/pools/pool-internal.ts:8]: @jerome-benoit Could you provide me with a resource to this 4.2 optimization? Maybe I missed something new :thinking:  

But @pioardi: this is only used internally and therefore it is okay as long it is not exported to public (e.g. via `index.ts`)

Comment (jerome-benoit) [reply src/pools/pool-internal.ts:8]: I do not follow closely all changes in TS release. All I can say here is that the transpiled code for that enum is: 

```js
exports.PoolType = void 0;
(function (PoolType) {
    PoolType["FIXED"] = "fixed";
    PoolType["DYNAMIC"] = "dynamic";
})(exports.PoolType || (exports.PoolType = {}));
```

which looks pretty efficient to me.

Comment (Shinigami92) [reply src/pools/pool-internal.ts:8]: `Object.freeze` doesn't get transpiled at all and are directly native JS  
Are JSDoc preserved on the enum values also, so JS users that don't use TS also see them?

Comment (Shinigami92) [reply src/pools/pool-internal.ts:8]: Used something to read I just quickly googled: https://blog.logrocket.com/why-typescript-enums-suck

tl;dr as long as it is just for internal usage it is okay IMO
But otherwise, please try to rely on more native JS
Users can then also use simply e.g. `'HARD'` instead of `KillBehaviors.HARD`

File: src/pools/abstract-pool.ts (hunk 21 above)
Comment (Shinigami92) [reply src/pools/abstract-pool.ts]: @jerome-benoit Did you see my last response here?

Comment (jerome-benoit) [reply src/pools/abstract-pool.ts]: I guess not :)

File: src/pools/pool-internal.ts (hunk 9 above)
Comment (jerome-benoit) [reply src/pools/pool-internal.ts:8]: Fair enough. 

So that makes that PR mergeable? ;-)

Comment (Shinigami92) [reply src/pools/pool-internal.ts:8]: From my side: Yes :+1:

**Comment by jerome-benoit** [comment]
I think that PR came to a mergeable agreement. So unless someone still have some disagreements, can one approve it and merge it? 

Thanks.

**Comment by Shinigami92** [comment]
I want to let @pioardi approve this :slightly_smiling_face:

**Review by pioardi — APPROVED** [review]

**Comment by jerome-benoit** [comment]
> I want to let @pioardi approve this 🙂

But that PR still needs that you approuve the requested changes you've asked ;)

**Review by Shinigami92 — APPROVED** [review]

**Comment by Shinigami92** [comment]
> But that PR still needs that you approve the requested changes you've asked ;)

GitLab is a bit better in smart detection for changes auto-resolval :thinking:
